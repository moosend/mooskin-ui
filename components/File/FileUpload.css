@import '../_utils/globals/variables.css';
@import '../_utils/globals/screens.css';

.fileContainer{
    display: flex;
    flex-grow: 1;
    flex-shrink: 1;
}

.contentContainer{
    display: flex;
    flex-grow: 1;
    flex-shrink: 1;
    position: relative;
}

.button{
    display: flex;
    justify-content: center;
    flex-direction: column;
    text-align: center;
    border-radius: 10px;
    /*font-size: 14px;*/
    /*padding: 10px 45;*/
    /* padding: 0.625rem 2.813rem; */
    font-size: .875rem;
    margin: 3px;
    background-color: var(--moosendBlue);
    border: 1px solid transparent;
    color: #ffffff;
    padding: 8px 38px;
}

.name{
    display: flex;
    flex-grow: 1;
    flex-shrink: 1;
    justify-content: center;
    flex-direction: column;
    padding: 15px;
    font-family: Arial,sans-serif;
    color: var(--inputFontColor);
    /*font-size: 16px*/
    font-size: 1rem;
    border-radius: 8px;
    background: #efefef;
}

.input{
    outline: none;
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    cursor: pointer;
    opacity: 0;
}

.disabledFile{
    cursor: not-allowed;
    opacity: 0.7;
}

.disabledFile input {
    cursor: not-allowed;
}

.label {
    display: flex;
    /* justify-content: center; */
    /* flex-direction: column; */
    color: var(--inputLabel);
    /* font-size: 13px; */
    font-size: .92857142857rem;
    flex-basis: 150px;
    padding-top: 15px;
}

.disabledFile > .name{
    cursor: not-allowed;
    color: #bfbfbf;
    background: #f6f6f6;
}

.disabledFile > .button{
    background: var(--radioHover);
}

.disabledFile > .input{
    cursor: not-allowed;
}

.description{
    display: block;
    color: var(--italicDescription);
    font-size: .78571428571rem;
    margin-top: 10px;
}

.success{
    box-shadow: inset 0px 0px 0px 2px #00cc44;
}

.error{
    box-shadow: inset 0px 0px 0px 2px #ec3232;
}

.descError{
    color: #ec3232;
}

.descSuccess{
    color: #00cc44;
}

@media (--xSmall){
    .fileContainer{flex-direction: column}
    .contentContainer{flex-direction: column}
    .label{
        margin-left: 3px;
        flex-basis: 20px;
    }

    .name{
        margin-bottom: 5px
    }

    .label {
        padding-top: 0;
    }

}
