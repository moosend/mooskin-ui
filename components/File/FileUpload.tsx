import * as React from 'react';

import styles from './FileUpload.css';

import {IInputCallbackData, IValidationCallbackData} from '../_utils/types/commonTypes';

export interface IFileProps{

    /** file upload id */
    id?: string;

    /** already selected file for upload */
    file?: File | FileList;

    /** wether the file upload field is required or not */
    required?: boolean;

    /** multiple file uploads */
    multiple?: boolean;

    /** what kind of file type does the file upload accept */
    accept?: string;

    /** file upload label */
    label?: string;

    /** with of the label within the fileupload container */
    labelWidth?: number;

    /** file upload input label */
    placeholder?: string;

    /** file upload button label */
    buttonLabel?: string;

    /** disabled file upload */
    disabled?: boolean;

    /** fileupload description (small italic bottom) */
    description?: string;

    /** status of the fileupload, error or success */
    status?: 'error' | 'success';

    /** validate function */
    validate?: (data: IValidationCallbackData) => boolean;

    /** override File styles */
    style?: React.CSSProperties;

    /** override File class */
    className?: string;

    /** what data is being used, helps whn extracting user input, you know on what field changes are made */
    dataLabel?: string;

    /** callback that is called when the input changes */
    onChange?: (e: React.ChangeEvent<HTMLInputElement>, data: IInputCallbackData) => void;
}

export interface IFileState{
    file?: File | FileList;
    name?: string;
}

export default class FileUpload extends React.Component<IFileProps, IFileState>{

    static defaultProps = {
        buttonLabel: 'Upload',
        className: '',
        placeholder: 'No File Selected...',
        style: {}
    };

    static displayName = 'FileUpload';

    static getDerivedStateFromProps(nextProps: IFileProps) {
        return {file: nextProps.file};
    }

    constructor(props: IFileProps){
        super(props);

        this.state = {
            file: this.props.file,
            name: this.props.file && this.fileNames(this.props.file) || this.props.placeholder
        };
    }

    // componentWillReceiveProps(nextProps: IFileProps){
    //     this.setState({file: nextProps.file});
    // }

    render(){

        const {disabled, description, accept, style, id, label, buttonLabel, required, multiple} = this.props;
        const disabledFile = this.props.disabled ? styles.disabledFile : '';
        const spacing = !this.props.labelWidth ? {} : {flexBasis: `${this.props.labelWidth}px`};
        const status = this.getStatus();
        const descStatus = this.getDescStatus();

        return (
            <div
                id={id}
                className={`file-upload-component ${styles.fileContainer} ${disabledFile}`}
                style={style}
            >
                {label && <label className={styles.label} style={spacing}>{label}</label>}
                <div style={{flexShrink: 1, flexGrow: 1, flexDirection: 'column'}}>
                    <div className={styles.contentContainer}>
                        <span className={`${styles.name} ${status}`}>{this.state.name}</span>
                        <span className={styles.button}>{buttonLabel}</span>
                        <input
                            accept={accept}
                            required={required}
                            disabled={disabled}
                            onChange={this.onChange}
                            type="file"
                            className={styles.input}
                            multiple={multiple}
                        />
                    </div>
                    {description && <i className={`${styles.description} ${descStatus}`}>{description}</i>}
                </div>
            </div>
        );
    }

    onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (e.target.files && e.target.files.length > 1){
            const files = e.target.files;
            const fileNames = this.fileNames(files);
            !this.props.disabled && this.setState({file: files, name: fileNames});
            !this.props.disabled && this.props.onChange &&
            this.props.onChange(e, {value: files, dataLabel: this.props.dataLabel});
            if (this.props.status){
                this.props.validate &&
                this.props.validate({value: files, dataLabel: this.props.dataLabel, required: this.props.required});
            }
        } else{
            const file = e.target.files && e.target.files[0];
            if (file){
                !this.props.disabled && this.setState({file, name: file.name});
                !this.props.disabled && this.props.onChange &&
                this.props.onChange(e, {value: file, dataLabel: this.props.dataLabel});
                if (this.props.status){
                    this.props.validate &&
                    this.props.validate({value: file, dataLabel: this.props.dataLabel, required: this.props.required});
                }
            }
        }
    }

    fileNames = (files: any) => {
        const fileNames = [];
        for (let i = 0 ; i < files.length ; i++){
            fileNames.push(files[i].name);
        }
        const fileStrings = fileNames.join(', ');
        return fileStrings;
    }

    getStatus = () => {
        const fileStatus = this.props.status && this.props.status;
        if (fileStatus){
            if (fileStatus === 'error'){
                return styles.error;
            } else if (fileStatus === 'success'){
                return styles.success;
            }
        }
    }

    getDescStatus = () => {
        const fileStatus = this.props.status && this.props.status;
        if (fileStatus){
            if (fileStatus === 'error'){
                return styles.descError;
            } else if (fileStatus === 'success'){
                return styles.descSuccess;
            }
        }
    }
}
