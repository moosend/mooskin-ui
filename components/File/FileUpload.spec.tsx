import toJson from 'enzyme-to-json';
import * as React from 'react';
import FileUpload from './FileUpload';

import { shallow } from 'enzyme';

describe('FileUpload', () => {

    test('renders correctly', () => {
        const func = jest.fn();

        const tree = shallow(
            <FileUpload
                onChange={func}
                disabled
                className="myClass"
                style={{color: 'blue'}}
                placeholder="input placeholder"
                label="with a label ofc"
                required
                buttonLabel="Upload"
                accept=".png"
                multiple
                id="5"
            />
        );
        expect(toJson(tree)).toMatchSnapshot();
    });

    test('renders properly into dom with props', () => {

        const component = shallow(
            <FileUpload accept=".png" multiple required placeholder="file" buttonLabel="upload"/>
        );

        expect(component.find('span').first().text()).toBe('file');
        expect(component.find('span').last().text()).toBe('upload');
        expect(component.find('input').prop('required')).toBeTruthy;
        expect(component.find('input').prop('miltiple')).toBeTruthy;
        expect(component.find('input').prop('disabled')).toBeFalsy;
        expect(component.find('input').prop('accept')).toEqual('.png');
    });

    test('onChange prop callback is called when a file is selected', () => {
        const func = jest.fn();

        const component = shallow(<FileUpload style={{color: 'blue'}} onChange={func}/>);

        component.find('input').simulate('change', { target: { files: [{name: 'File1'}] }});

        expect(func).toHaveBeenCalled();
        expect(component.find('div').prop('style')).toEqual({color: 'blue'});
        expect(component.find('span').first().text()).toBe('File1');
    });

    test('onChange prop callback is called when multiple files are selected', () => {
        const func = jest.fn();

        const component = shallow(<FileUpload className="myClass" multiple onChange={func}/>);

        component.find('input').simulate('change', { target: { files: [{name: 'File1'}, {name: 'File2'}] }});

        expect(func).toHaveBeenCalled();
        expect(component.find('div').hasClass('myClass')).toBeTruthy;
        expect(component.find('span').first().text()).toBe('File1, File2');
    });

    test('onChange prop callback is not called when disabled prop has been passed', () => {
        const func = jest.fn();

        const component = shallow(<FileUpload disabled onChange={func}/>);

        component.find('input').simulate('change', { target: { files: [{name: 'File1'}] }});

        expect(func).not.toHaveBeenCalled();
        expect(component.find('span').first().text()).toBe('No File Selected...');
    });

});
