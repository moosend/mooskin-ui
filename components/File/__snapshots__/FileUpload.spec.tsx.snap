// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload renders correctly 1`] = `
<div
  className="file-upload-component fileContainer disabledFile"
  id="5"
  style={
    Object {
      "color": "blue",
    }
  }
>
  <label
    className="label"
    style={Object {}}
  >
    with a label ofc
  </label>
  <div
    style={
      Object {
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <div
      className="contentContainer"
    >
      <span
        className="name undefined"
      >
        input placeholder
      </span>
      <span
        className="button"
      >
        Upload
      </span>
      <input
        accept=".png"
        className="input"
        disabled={true}
        multiple={true}
        onChange={[Function]}
        required={true}
        type="file"
      />
    </div>
  </div>
</div>
`;
